### ISSUE TITLE
Lots of test flake when upgrading from 4.7.0 to 4.8.0

### ISSUE THREAD

**Issue by mockdeep** [dequelabs/axe-core#4244]
### Product

axe-core

### Product Version

4.8.0

### Latest Version

- [X] I have tested the issue with the latest version of the product

### Issue Description

This is a repost from [this issue](https://github.com/dequelabs/axe-core-gems/issues/349) on the gems repo. I suspect it has more to do with the underlying `axe-core` library, though, than the gem.

#### Expectation

I expect running `axe` in our automated tests to consistently pass on pages in our CI environment.

#### Actual

Even when logging in our users in our end-to-end tests, `axe-core` will occasionally report accessibility violations like this:

```
Found 2 accessibility violations:

1) landmark-one-main: Document should have one main landmark (moderate)
    https://dequeuniversity.com/rules/axe/4.8/landmark-one-main?application=axeAPI
    The following 1 node violate this rule:
    
        Selector: html
        HTML: <html><head></head><body></body></html>
        Fix all of the following:
        - Document does not have a main landmark
        
2) page-has-heading-one: Page should contain a level-one heading (moderate)
    https://dequeuniversity.com/rules/axe/4.8/page-has-heading-one?application=axeAPI
    The following 1 node violate this rule:
    
        Selector: html
        HTML: <html><head></head><body></body></html>
        Fix all of the following:
        - Page must have a level-one heading
        
Invocation: axe.run({:exclude=>[]}, {:rules=>{"document-title"=>{:enabled=>false}, "html-has-lang"=>{:enabled=>false}}}, callback);
```

However, it looks like axe-core is checking against a blank page, even though we've asserted that the page is loaded. These are server-rendered Rails pages without any client-rendering. There's a bunch of other JS loading on the page, but it doesn't do much on the home screen.

There was a little flake on 4.7.0, but we found that we could auto-retry the a11y checks after a couple of seconds and they would work. Unfortunately, the retry logic doesn't resolve the issue on 4.8.0. It retries several times and still fails.

#### How to Reproduce

This is pretty hard since it doesn't happen consistently. I would say something like 10% of builds fails with these errors. Happy to hop on a screen share and walk through it in more detail if that would help.

### Additional Context

We've gone so far as to add extra expectations to ensure the page is loaded and still see this failure:

```
expect(page).to have_css('div')
expect(page).to be_axe_clean
```

**Comment by mockdeep** [comment]
@straker thanks for your response. We've got an assertion on the page to ensure that it's fully loaded. For example, the issue appears when logging in a user in our specs:

```rb
page.click_button('LOG IN')
expect(page).to have_content('MANAGE')
expect(page).to be_axe_clean
```

Where `'MANAGE'` only appears on the home page after the user is logged in. We've introduced some wrapper logic to retry it with a back-off, so if it fails immediately, we wait 2 seconds and try again, then 4, then 8, and eventually fail the test after 32 seconds.

I'll continue to tinker with things on our end, but one thing to note is we didn't notice this kind of flake on version 4.7.0. It seems to have started when we upgraded to 4.8.0.

**Comment by WilcoFiers** [comment]
Ensuring the page is in the correct state (and remains in that state) is not something axe-core can be responsible for. I'm afraid this is something you're going to have to solve in a different way. I don't know why axe-core 4.8 would have changed your situation. We turned off a few relatively slow rules in 4.8, so perhaps the timing changed just enough that a race condition that wasn't giving you problems before now is. Either way the solution is to address the race condition. Different testing frameworks have different solutions for it. A common one is to wait for a particular element to get loaded before doing a follow-up action.

**Comment by mockdeep** [comment]
@WilcoFiers this is a synchronous, server-rendered page. `expect(page).to have_content('MANAGE')` ensures it's fully rendered. There is no asynchronous rendering on the page, no React or anything else.

This seems to happen most often early in the test while logging in the user, and less so later in the spec. One thing I'm wondering is whether calling `be_axe_clean` in rapid succession might cause something like this. We call it on every page visit and button click. Is there any caching or asynchronous behavior in axe that might have a lingering affect? For more context, our login method looks like this:

```rb
visit(signin_path)
expect(page).to have_button('CONTINUE')
expect(page).to be_axe_clean
fill_in('session_email', with: user.email)

click_button('CONTINUE')
expect(page).to have_button('LOG IN')
expect(page).to be_axe_clean
fill_in('session_password', with: user.password)

click_button('LOG IN')
expect(page).to have_content('MANAGE')
expect(page).to be_axe_clean
```

What's odd is that we see the failure most frequently on the 3rd `be_axe_clean` call. Is it possible there is some lingering logic or caching from the previous calls?
